Store.js

```diff
@@ -140,14 +151,168 @@ define(["dcl/dcl", "decor/Invalidating"], function (dcl, Invalidating) {
+							this._hasItemObservable = true;
+							this._observeResults =  [];
+							// affect the callback to the observe function if the item is observable
+							this._observeResults[i + 1] = Observable.observe(this.store[i],
+								this.observeCallback.bind(this));
+						}
+					}
+					collection = processQueryResult.call(this, this.store.filter(this._isQueried, this));
+					if (this._isArrayObservable(this.store)) {
+						if (!this._observeResults) {
+							this._observeResults = [];
+						}
+						// affect the callback to the observe function if the array is observable
+						this._observeResults[0] = ObservableArray.observe(this.store, this.observeCallback.bind(this));
+					}
 				}
+				return this.processCollection(collection);
 			} else {
 				this.initItems([]);
 			}
 		},
 
 		/**
+		 * Synchronously deliver change records to all listeners registered via `observe()`.
+		 */
+		deliver: dcl.superCall(function (sup) {
+			return function () {
+				sup.call();
+				if (this._observeResults) {
+					this._observeResults[0].deliver();
+					if (this._hasItemObservable) {
+						this.observeCallback && Observable.deliverChangeRecords(this.observeCallback);
+					}
+				}
+			};
+		}),
+
+		/**
+		 * Discard change records for all listeners registered via `observe()`.
+		 */
+		discardChanges: dcl.superCall(function (sup) {
+			return function () {
+				sup.call();
+				if (this._observeResults) {
+					this._beingDiscarded = true;
+					this._observeResults[0].deliver();
+					if (this._hasItemObservable === true) {
+						this.observeCallback && Observable.deliverChangeRecords(this.observeCallback);
+					}
+					this._beingDiscarded = false;
+					return this.store;
+				}
+			};
+		}),
+
+		/**
+		 * Called to verify if an item is observable
+		 * @param item
+		 * @private
+		 */
+		_isItemObservable: function (item) {
+			return (Observable.test(item) || has("object-observe-api"));
+		},
+
+		/**
+		 * Called to verify if the array is observable
+		 * @param item
+		 * @private
+		 */
+		_isArrayObservable: function (item) {
+			return (ObservableArray.test(item) || has("object-observe-api"));
+		},
+
+		/**
+		 * Called to verify if the item respect the query conditions
+		 * @param item
+		 * @private
+		 */
+		_isQueried: function (item) {
+			for (var prop in this.query) {
+				if (item[prop] !== this.query[prop]) {
+					return false;
+				}
+			}
+			return true;
+		},
+
+		/**
+		 * Called when a modification is done on the array or its items.
+		 * @param changeRecords - send by the Observe function
+		 */
+		observeCallback: function (changeRecords) {
+			if (!this._beingDiscarded) {
+				for (var i = 0; i < changeRecords.length; i++) {
+					// array modified
+					if (Array.isArray(changeRecords[i].object)) {
+						this.observeCallbackArray(changeRecords[i]);
+						// one item of the array modified
+					} else {
+						var found = false;
+						for (var j = i + 1; j < changeRecords.length; j++) {
```

Comment:
Rather change doing two-dimensional array here,  you may do:

```javascript
var objects = [];
for (var i = 0; i < changeRecords.length; ++i) {
    var object = changeRecords[i].object;
    if (objects.indexOf(object) < 0) {
        objects.push(object);
        callback(object);
    }
}
```

Note that it just sends the object to the callback, given we only care about an object is changed or not.

The logic will be even simpler if we make sure only one object is handled in a callback, in such case it’d be like `callback(changeRecords[0].object)`.